libr/io/map.c

```diff
@@ -1,372 +1,265 @@
-/* radare - LGPL - Copyright 2008-2016 - pancake */
+/* radare2 - LGPL - Copyright 2017 - condret */
 
-#include <stdio.h>
-#include <string.h>
-#include <stdlib.h>
 #include <r_io.h>
 #include <r_util.h>
-#include <r_list.h>
+#include <sdb.h>
+#include <stdlib.h>
 
-R_API int r_io_map_count (RIO *io) {
-	return ls_length (io->maps);
-}
+#define END_OF_MAP_IDS  0xffffffff
 
-R_API RIOMap * r_io_map_new(RIO *io, int fd, int flags, ut64 delta, ut64 addr, ut64 size) {
-	RIOMap *map = R_NEW0 (RIOMap);
-	if (!map || ((UT64_MAX - size) < addr)) { //prevent interger-overflow
+R_API RIOMap* r_io_map_new(RIO* io, int fd, int flags, ut64 delta, ut64 addr, ut64 size) {
+	RIOMap* map = NULL;
+	if (!size || !io || !io->maps || ((UT64_MAX - size + 1) < addr) || !io->map_ids) {
+		return NULL;
+	}
+	map = R_NEW0 (RIOMap);
+	if (!map || !io->map_ids || !r_id_pool_grab_id (io->map_ids, &map->id)) {
 		free (map);
 		return NULL;
 	}
 	map->fd = fd;
+	map->from = addr;
+	// RIOMap describes an interval of addresses (map->from; map->to)
+	map->to = addr + size - 1;
```

Comment:
```diff
-	map->to = addr + size - 1;
+	map->to = addr + size;
```

I think 2^64-1 address should not be considered to support in r2. So don't worry about address wraparound.

https://github.com/radare/radare2/pull/8239